### DISCUSSION TITLE
Crystal Plasticity with Grain Tracker

### DISCUSSION THREAD

**Discussion by FHilty** [idaholab/moose#21484] [Q&A General] [answered]
Me and @jbair34 would like to combine crystal plasticity with the grain trackers capabilities for discrete nucleation. However, we are stuck as the ComputeElasticityTensorCP material is needed to create the necessary material properties for the crystal plasticity kernels, but there isn't a "grain_tracker" parameter for this material as there is with other ComputeElasticityTensor options.  As a result the crystal plasticity kernels isn't interacting with the grain structure from grain tracker. Is there currently a way to make this work?

**Comment by GiudGiud** [comment]
@permcody @dschwen who else may I tag on grain tracking

**Reply by FHilty** [reply]
Just poking this thread to see if there are any updates.

**Reply by permcody** [reply]
The GrainTracker can certainly be added to that material. I don’t have any resources that can work on this but we can advise somebody who would be willing to do the work.

**Reply by FHilty** [reply]
Hi Cody, thank you for the information. In a few week I will have the time to work on the changes, I will touch base again when I can get started. Thanks again.

**Reply by FHilty** [reply]
My schedule cleared up to where i can work on this and have been reading through the crystal plasticity and grain tracker files. 

There are three material objects used `ComputeElasticityTensorCP`, `ComputeMultipleCrystalPlasticityStress`, and (in our case) `CrystalPlasticityKalidindiUpdate`. I don't believe the latter two will need modification, however they do utilize the material property `crysrot` that is provided from `ComputeElasticityTensorCP` so that is the first object that I will need to modify. From looking at the material `ComputePolycrystalElasticityTensor` I have a fair idea of how to give a material access to the grain tracker and retrieve the correct grain ID at an element. 

My first question is what does the command `_grain_tracker.getData(grain_id)` actually return? That exact command seems to be used to retrieve both the Euler angles and elasticity tensor in different places. I think modifying `ComputeElasticityTensorCP` to (optionally) utilize the grain tracker will be straight forward once I understand how to get those specific pieces of information. This update would mean that the correct `crysrot` material property is provided to the other material objects.

My second question is much more open ended, what other objects would need modification to couple crystal plasticity with the grain tracker? I have not seen any other places in the kernel/material objects I need which require access grain trackers information.


____
Reduced version of the input script (extension changed from .i to .txt):
[GT_DN_XP_Simple.txt](https://github.com/idaholab/moose/files/9253408/GT_DN_XP_Simple.txt)

EBSD data for structure (extension changed from .ebsd to .txt):
[MCPC_20Grain_ebsd.txt](https://github.com/idaholab/moose/files/9253421/MCPC_20Grain_ebsd.txt)

Euler angle file that has correct inital 20 grain orientations and random angles for nucleated grains:
[MCPC_Random_Euler_1000.txt](https://github.com/idaholab/moose/files/9253422/MCPC_Random_Euler_1000.txt)

Slip system text file:
[input_slip_sys.txt](https://github.com/idaholab/moose/files/9253447/input_slip_sys.txt)

**Comment by permcody** [comment, marked as answer]
First Answer: `.getData(grain_id)` is a templated interface function. It can return just about anything we define. It's setup to return various attributes for a given grain (things like Euler Angles, volumes, proximity to the boundary, etc). To see what methods are available you'll need to dig through the `functor` interface classes in the PhaseField module. I'm most familiar with `EBSDAccessorFunctors`, but as you've seen there are others. Just look at what interfaces the class you are interested in inherits from and look at the associated methods in those classes. 

Second Answer: The whole point of the GrainTracker is to maintain a stable ID with a reduced number of order parameters so you'll likely need several attributes of each Grain to perform your calculations and this is where the GrainTracker can help you. I'm fairly certain that we are pulling unique grain information in the material system already, which feeds into the residual calculations for some of these models.

**Reply by FHilty** [reply]
Thank you for the assistance and sorry for the long silence, I had not heard of functors before so it was a learning experience. Once I started to understand what exactly was happening in other MOOSE objects I found that the solution was much simpler than I thought.

For anyone finding this thread the solution I settled on was to start with the `ComputeElasticityTensorCP` material and replaced the body of `assignEulerAngles()` with 2 steps. First retrieve the grain ID from grain tracker, then look up the Euler angles directly from the file using the Euler angle provider. As it turns out, this is exactly what `OutputEulerAngles` does in the `precalculateValue()` function. I performed a few tests and, so far as I can tell, the rest of the crystal plasticity framework is working correctly, it just hinged on getting the correct Euler angles. Now when the grain tracker inserts a new grain nucleus with discrete nucleation it is reflected in the stress/strain fields from the crystal plasticity modules.

**Reply by permcody** [reply]
Grea, glad you are working! The Euler Angle provider caches values so it doesn't have to do a file lookup each time you call it. There is a lot of nifty, but complex machinery making all this work. Let us know if you need any further assistance.
